File: Arrays/53-Maximum-Subarray/Python/53-Maximum-Subarray-Better.py
"""
Problem: 53. Maximum Subarray
Approach: Better - Prefix Sum

Time Complexity: O(n^2)
Space Complexity: O(n)
"""


class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        n = len(nums)

        prefix = [0] * (n + 1)

        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]

        maxi = float("-inf")

        for i in range(n):
            for j in range(i, n):
                total = prefix[j + 1] - prefix[i]
                maxi = max(maxi, total)

        return maxi

File: Arrays/53-Maximum-Subarray/JavaScript/53-Maximum-Subarray-Better.js
/**
 * Problem: 53. Maximum Subarray
 * Approach: Better - Prefix Sum
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */

var maxSubArray = function(nums) {
    const n = nums.length;

    const prefix = new Array(n + 1).fill(0);

    for (let i = 0; i < n; i++) {
        prefix[i + 1] = prefix[i] + nums[i];
    }

    let maxi = -Infinity;

    for (let i = 0; i < n; i++) {
        for (let j = i; j < n; j++) {
            const total = prefix[j + 1] - prefix[i];
            maxi = Math.max(maxi, total);
        }
    }

    return maxi;
};